### PR TITLE
Refactor add_tags method

### PR DESCRIPTION
**Pull request by gabegma**
Part of #297

## Description:
* Refactor `add_tags` method so it is cleaner. it seems some code was redundant and that we were saving prediction tags in the base dataset split.
* This triggered many `mypy` errors so I needed to refactor further the types related to tags.
* This is preliminary work before importing proposed actions.

## Checklist:

You should check all boxes before the PR is ready. If a box does not apply, check it to acknowledge it.

* [x] **ISSUE NUMBER.** You linked the issue number (Ex: Resolve #XXX).
* [x] **PRE-COMMIT.** You ran pre-commit on all commits, or else, you
  ran `pre-commit run --all-files` at the end.
* [x] **USER CHANGES.** The changes are added to CHANGELOG.md and the documentation, if they impact
  our users.
* [x] **DEV CHANGES.**
    * Update the documentation if this PR changes how to develop/launch on the app.
    * Update the `README` files and our wiki for any big design decisions, if relevant.
    * Add unit tests, docstrings, typing and comments for complex sections.

### CODE REVIEW

File: azimuth/dataset_split_manager.py [hunk 1]
```diff
@@ -261,79 +261,45 @@ def _init_dataset_split(self, dataset_split: Dataset, tags) -> Dataset:
...
-        """
-
-        def check_tags(tags_dict):
-            diff_key = set(tags_dict.keys()).difference(self._tags + self._prediction_tags)
-            if len(diff_key) > 0:
-                raise ValueError(f"Unknown tags: {diff_key}. Expected one of {self._tags}")
-            return tags_dict
-
-        ds = ds.map(lambda u, i: check_tags(tags.get(i, {})), with_indices=True, desc="Set Tag")
-        self._save_base_dataset_split()
```
Comment (gabegma) [inline azimuth/dataset_split_manager.py:281]: @Dref360 - it seems we were saving the prediction tags in `self._save_base_dataset_split()`, and saving twice `base_dataset_split` for base tags. Am I missing something with this refactoring?

Comment (Dref360) [reply azimuth/dataset_split_manager.py:281]: I think the `self._save_base_dataset_split()` was not useful on line 281.
The rest of the refactor looks good.

Comment (gabegma) [reply azimuth/dataset_split_manager.py:281]: Thanks for validating, that was our intuition too.

**Review by JosephMarinier — APPROVED** [review]
Wow! 😲 Good job! 👌 Thank you for cleaning that up.

**Pull request merged**